lib/Sema/CSApply.cpp

```diff
@@ -7295,39 +7302,104 @@ static Expr *finishApplyDynamicCallable(ConstraintSystem &cs,
 
   // Construct expression referencing the `dynamicallyCall` method.
-  Expr *member =
+  MemberRefExpr *member =
     new (ctx) MemberRefExpr(fn, fn->getEndLoc(), ConcreteDeclRef(method),
                             DeclNameLoc(method->getNameLoc()),
                             /*Implicit*/ true);
 
+  cs.setType(member, methodType);
+
+  bool isDynamic
+      = selected.choice.getKind() == OverloadChoiceKind::DeclViaDynamic;
+  Expr* memberExpr = buildMemberRef(
+      member->getBase(), selected.openedFullType, member->getDotLoc(),
+      selected.choice, member->getNameLoc(), selected.openedType,
+      cs.getConstraintLocator(member), loc, true,
+      selected.choice.getFunctionRefKind(), member->getAccessSemantics(),
+      isDynamic);
+
   // Construct argument to the method (either an array or dictionary
   // expression).
   Expr *argument = nullptr;
   if (!useKwargsMethod) {
-    argument = ArrayExpr::create(ctx, SourceLoc(), arg->getElements(),
+    auto paramArrayType = methodType->getParams()[0].getParameterType();
+    auto paramType = dyn_cast<ArraySliceType>(paramArrayType.getPointer())->getBaseType();
+
+    std::vector<Expr*> convertedElements;
+    for (Expr* dynamicArg: arg->getElements()) {
+      convertedElements.push_back(
+          coerceToType(dynamicArg, paramType,
+                                               loc.withPathElement(
+             LocatorPathElt::getApplyArgToParam(0, 0)))
+          );
+    }
+    auto* argumentTmp = ArrayExpr::create(ctx, SourceLoc(), convertedElements,
                                  {}, SourceLoc());
+    argument = argumentTmp;
+
+    cs.setType(argument, paramArrayType);
+    finishArrayExpr(argumentTmp);
   } else {
+    // Extract expected typle types (Key, Value) from the protocol conformance.
+    auto paramDictType = methodType->getParams()[0].getParameterType();
+    auto dictLitProto =
+      cs.TC.getProtocol(fn->getLoc(), KnownProtocolKind::ExpressibleByDictionaryLiteral);
+    auto conformance = cs.TC.conformsToProtocol(
+        paramDictType, dictLitProto, cs.DC, ConformanceCheckFlags::InExpression);
+    auto keyAssocType =
+        ProtocolConformanceRef::getTypeWitnessByName(
+            paramDictType, *conformance, cs.getASTContext().Id_Key, &cs.TC)
+            ->getDesugaredType();
+    auto valueAssocType =
+        ProtocolConformanceRef::getTypeWitnessByName(
+            paramDictType, *conformance, cs.getASTContext().Id_Value, &cs.TC)
+            ->getDesugaredType();
+
     SmallVector<Identifier, 4> names;
     SmallVector<Expr *, 4> dictElements;
     for (unsigned i = 0, n = arg->getNumElements(); i < n; i++) {
       Expr *labelExpr =
         new (ctx) StringLiteralExpr(arg->getElementName(i).get(),
                                     arg->getElementNameLoc(i),
                                     /*Implicit*/ true);
+      (void)cs.TC.typeCheckExpression(labelExpr, dc, TypeLoc::withoutLoc(keyAssocType),
```

Comment:
Please find a way to remove this typeCheckExpression() call as well

Reply:
This was necessary because it was requiring the type to be registered with the cs solution. The check was added in 2016 but seems like it was made unnecessary in 2017 because simplifyType became an identity for such types. I just removed the check. I hope that is ok.